### DISCUSSION TITLE
Why only records?

### DISCUSSION THREAD

**Discussion by javafanboy** [jerolba/parquet-carpet#40] [General]
What was the thinking behind only supporting records for the serialization in Carpet?

I do right now for instance have a use-case where I receive binary data objects over a network and each record have large number of non-mandatory parameters (a bit mask is used to specify the included ones that are then  deserialize one by one).  For this I now need to create and maintain a class (or separate local variables in my processing class) that is exactly identical to the record I use for creating the Parquet file BUT HAVE MUTABLE ATTRIBUTES that I set one by one as they are read and when complete turned into the record required by Carpet. 

Sadly I also need to create a NEW record for each object (I will receive hundreds of millions) rather than being able to re-use a single object that is reset each time. This creates a lot of GC pressure.

There are probably others with similar needs for "mutability" and everyone producing huge files would probably like to improve performance by re-using the serialization record - i.e. why not supporting BOTH records and normal classes?

**Comment by jerolba** [comment]
Because records are simple, I want to keep the library as simple as possible. :)

The data model proposed by Java Records is very straightforward: it provides a canonical constructor for deserialization, and the serialization of a record is based solely on its state components.

Serialization and deserialization can be quite messy... What makes an attribute _serializable_? How can I access it? Through a setter/getter? What happens if it's not accessible, or if it's final? You start implementing hacks for the sake of simplicity, but eventually, you run into security issues and code full of conditions.

I encourage you to read some articles related to serialization and security:
- [Towards Better Serialization](https://openjdk.org/projects/amber/design-notes/towards-better-serialization) by @briangoetz 
- [Simpler object and data serialization using Java records](https://blogs.oracle.com/javamagazine/post/simpler-object-and-data-serialization-using-java-records)
- [Java JSON deserialization problems with the Jackson ObjectMapper](https://snyk.io/blog/java-json-deserialization-problems-jackson-objectmapper/)

I don't want to dive into the complexities and use cases that _normal_ Classes require, like Jackson.

The value proposition of Carpet is to use it in Data Eng. domain with a lot of data, but not Big Data, endorsing [Data-Oriented Programming patterns](https://www.infoq.com/articles/data-oriented-programming-java/), where records play a key role. Even though there may be many instantiations, the overhead of mapping a Class to/from a Record is negligible compared to the overall process time. Performance is considered when possible, but it's not the first priority. 

For your reassurance, I am currently working on a feature in Carpet to abstract the "data model," making it possible to avoid the use of records in exchange for some configurable mapping.

This will be an example of usage:

```java
var mapper = recordOf(YourClass.class)
        .withColumn("id", LONG.notNull(), YourClass::id)
        .withColumn("code", INTEGER, YourClass::code)
        .withColumn("name", STRING, YourClass::name)
        .withColumn("foo", STRING, o -> o.myMap.get("foo"))
        .withColumn("bar", STRING, o -> o.myMap.get("bar"))
        .withColumn("interest", DOUBLE, o -> o.amount * o.number *  0.13)
        .withColumn("amount", DOUBLE, o -> o.amount * o.number);

try(var writer = CarpetWriter.builder(output, YourClass.class).withRecordModel(mapper).build()) {
    //write your data
}
```

I don't know when it will be ready for a release. 

To solve your problem, in the meantime, you can try [JFleet Parquet](https://github.com/jerolba/jfleet/tree/master/jfleet-parquet), other library that I maintain that follow this pattern to write into Postgres/MySql tables and CSV/Avro/Parquet files.

Dependency:

```xml
<dependency>
    <groupId>org.jfleet</groupId>
    <artifactId>jfleet-parquet</artifactId>
    <version>0.6.8</version>
</dependency>
```

[Code sample](https://github.com/jerolba/jfleet/blob/master/jfleet-parquet/src/test/java/org/jfleet/parquet/JFleetParquetWriterTest.java#L380):

```java
EntityInfo entityInfo = new EntityInfoBuilder<>(TestEntity.class)
      .addColumn("fooBoolean", BOOLEAN, TestEntity::getFooBoolean)
      .addColumn("fooInt", INT, TestEntity::getFooInt)
      .addColumn("fooShort", SHORT, TestEntity::getFooShort)
      .addColumn("fooByte", BYTE, TestEntity::getFooByte)
      .addColumn("fooDouble", DOUBLE, TestEntity::getFooDouble)
      .addColumn("fooLong", LONG, TestEntity::getFooLong)
      .addColumn("fooFloat", FLOAT, TestEntity::getFooFloat)
      .build();

try (FileOutputStream outputStream = new FileOutputStream(testFile)) {
    ParquetConfiguration<TestEntity> parquetConfiguration = new ParquetConfiguration.Builder<TestEntity>(outputStream, entityInfo).build();
    try (JFleetParquetWriter<T> parquetWriter = new JFleetParquetWriter<>(parquetConfiguration)) {
        while (condition)
                 parquetWriter.write(getNewEntity());
        }
}
```

You can reuse the entity, without creating one in each iteration.

Drawbacks of JFleet Parquet:
* Supports fewer types
* Can only write plain tables, no nested data structures
* Uses Parquet 1.13.1

If you finally try it, can you measure the performance difference? How much do you save if you don't create the records or even the entity class?

**Comment by javafanboy** [comment]
Thanks for the explanation and proposals!

I agree that serialization is a tricky subject (have read all the articles
you linked before). In my use case I do not REALLY need full blown object
serialization as I do not recreate a specific object instance and have a
very static set of data in mind that is not likely to change much in the
future.

Today I start with a "stream" of object instances in my clients but from
there for each just sends a byte mask indicating what is included and then
serialize the raw data in a determined order (variable length encoded) with
zero object headers etc. On the receiving side I, to be able to keep as
many records in memory as possible at a time given available RAM, keep the
data in this compact byte array form through out aggregation and even some
sorting (with a custom comparator the read run length encoded numeric
types) all the away until I need to generate the parquet data at which
point I unpack one record at the time (reading the bitmask and reading the
data in the same order as written) into a reused mutable object instance
(not the same type as I started with but rather a simple wrapper designed
specifically for this purpose) one attribute at a time as they are decoded.
To use Carpet I right now then creates a record from the reused object
(letting all null-values go straight trough) that I pass in to the Carpet
writer and then immediately throw away. The carpet writer will directly
stream the data (I assume one row-group at a time) to a custom stream that
uploads data to an S3 object (that I keep writing to until it reaches an
optimal size of a few hundred MB or a max update interval is reached) at
which point a new S3 object is created along with a new stream and carpet
writer.
I believe this still can be reasonably efficient (java is good at garbage
collecting short lived object) but I am looking for all possible
improvements (many small ones together tend to make a big difference) to be
able to keep up with an as high data rate as possible (my communication
channel is asynchronous so I will not loose any data if I in "spikes" cant
read fast enough but the steady state performance need to bee as good as I
can make it).

As for Carpet as it is today I am very happy with the library in general
including the excellent support for structured attributes, that is quite
complex to program with low-level libraries, that I leverage to allow for
more efficient queries by engines that "understand" this on the Parquet
level (I will use AWS Athena that do support it) and to not having to
bother with "quoting" etc. for nested elements that I would need to do if I
myself expanded say lists of strings to CSV or some other text format and
stored as a single string.

Your new features, when they are ready, could be very useful for me, in
particular if it supported ONE level of structured types (i.e. Java
collections with only non-structured objects like strings or Integers but
not collections of collections).

For now I have almost implemented the whole aggregator the best I can with
the Carpet library as it is (as described above) and will evaluate
performance. If good enough I will stay with that until you provide any
updated version or I find some more suitable library etc.
